### ISSUE TITLE
Accelerate Allreduce of IndexedSlices via Compression

### ISSUE THREAD

**Issue by gibiansky** [horovod/horovod#1481]
**Is your feature request related to a problem? Please describe.**

When using `tf.gather` or `tf.nn.embedding_lookup`, the gradient is a `tf.IndexedSlices` object. Reducing these `IndexedSlices` can be slow.

**Describe the solution you'd like**

Apply FP16 gradient compression to `IndexedSlices`. Right now, [compression is only applied to dense tensors](https://github.com/horovod/horovod/blob/master/horovod/tensorflow/__init__.py#L68). 

Does this seem like a reasonable change? If so, I can make a PR for it.

**Comment by gibiansky** [comment]
I think this is noj
